File: sonara-industries/apps/web/lib/formulas.ts
const EPSILON = 1e-6;

export function clamp(value: number, low = 0, high = 100) {
  return Math.max(low, Math.min(high, value));
}

export function geometricWeightedScore(scores: Record<string, number>, weights: Record<string, number>) {
  const totalWeight = Object.values(weights).reduce((sum, weight) => sum + Math.max(weight, 0), 0);
  if (totalWeight <= 0) return 0;
  const factors = Object.entries(scores)
    .map(([key, score]) => {
      const weight = Math.max(weights[key] ?? 0, 0);
      if (weight === 0) return null;
      return Math.max(clamp(score) / 100, EPSILON) ** (weight / totalWeight);
    })
    .filter((value): value is number => value !== null);
  if (factors.length === 0) return 0;
  return Number((clamp(factors.reduce((product, value) => product * value, 1) * 100)).toFixed(2));
}

export function readinessScore(checks: Record<string, number>) {
  return geometricWeightedScore(checks, {
    metadata: 1.2,
    rights: 1.4,
    assets: 1,
    strategy: 0.9,
    quality: 1.1,
    approval: 1.5,
  });
}

export function laborPercentage(totalLaborCost: number, projectedSales: number) {
  if (projectedSales <= 0) return 0;
  return Number(((totalLaborCost / projectedSales) * 100).toFixed(2));
}

export function recipeCost(ingredientCosts: number[], servings: number) {
  if (servings <= 0) return 0;
  return Number((ingredientCosts.reduce((sum, cost) => sum + Math.max(cost, 0), 0) / servings).toFixed(2));
}

export function menuTargetPrice(costPerServing: number, targetFoodCostPercent: number) {
  if (targetFoodCostPercent <= 0) return 0;
  return Number((costPerServing / (targetFoodCostPercent / 100)).toFixed(2));
}

export function alertSeverity(hazard: number, proximity: number, urgency: number, sourceTrust: number) {
  return Number(clamp(hazard * 0.35 + proximity * 0.2 + urgency * 0.3 + sourceTrust * 0.15).toFixed(2));
}

export function securityRisk(sensitivity: number, exposure: number, automationPower: number, sourceTrust: number) {
  return Number(clamp(sensitivity * 0.3 + exposure * 0.25 + automationPower * 0.25 + (100 - clamp(sourceTrust)) * 0.2).toFixed(2));
}

export function queueUtilization(arrivalRate: number, serviceRate: number) {
  if (serviceRate <= 0) return 1;
  return Number((arrivalRate / serviceRate).toFixed(4));
}

export function exponentialSmoothing(previousForecast: number, actual: number, alpha = 0.35) {
  const boundedAlpha = Math.max(0, Math.min(1, alpha));
  return Number((boundedAlpha * actual + (1 - boundedAlpha) * previousForecast).toFixed(4));
}

export function moatScore(inputs: {
  networkEffects: number;
  switchingCost: number;
  proprietaryData: number;
  brand: number;
  compliance: number;
  executionSpeed: number;
}) {
  return Number(
    clamp(
      inputs.networkEffects * 0.18
        + inputs.switchingCost * 0.2
        + inputs.proprietaryData * 0.2
        + inputs.brand * 0.14
        + inputs.compliance * 0.16
        + inputs.executionSpeed * 0.12,
    ).toFixed(2),
  );
}

export function pricingScore(costSavings: number, timeSavings: number, urgency: number, segmentFit: number) {
  return Number(clamp(costSavings * 0.3 + timeSavings * 0.25 + urgency * 0.2 + segmentFit * 0.25).toFixed(2));
}

File: sonara-industries/apps/api/app/formulas.py
from __future__ import annotations

from math import prod

EPSILON = 1e-6


def clamp(value: float, low: float = 0.0, high: float = 100.0) -> float:
    return max(low, min(high, value))


def geometric_weighted_score(scores: dict[str, float], weights: dict[str, float]) -> float:
    """Geometric weighted score, protected from zero collapse by EPSILON."""
    total_weight = sum(max(weight, 0.0) for weight in weights.values())
    if total_weight <= 0:
        return 0.0
    factors: list[float] = []
    for key, score in scores.items():
        weight = max(weights.get(key, 0.0), 0.0)
        if weight == 0:
            continue
        normalized = clamp(score, 0.0, 100.0) / 100.0
        factors.append((max(normalized, EPSILON)) ** (weight / total_weight))
    if not factors:
        return 0.0
    return round(clamp(prod(factors) * 100.0), 2)


def readiness_score(checks: dict[str, float]) -> float:
    weights = {
        "metadata": 1.2,
        "rights": 1.4,
        "assets": 1.0,
        "strategy": 0.9,
        "quality": 1.1,
        "approval": 1.5,
    }
    return geometric_weighted_score(checks, weights)


def labor_percentage(total_labor_cost: float, projected_sales: float) -> float:
    if projected_sales <= 0:
        return 0.0
    return round((total_labor_cost / projected_sales) * 100.0, 2)


def recipe_cost(ingredient_costs: list[float], servings: float) -> float:
    if servings <= 0:
        return 0.0
    return round(sum(max(cost, 0.0) for cost in ingredient_costs) / servings, 2)


def menu_target_price(cost_per_serving: float, target_food_cost_percent: float) -> float:
    if target_food_cost_percent <= 0:
        return 0.0
    return round(cost_per_serving / (target_food_cost_percent / 100.0), 2)


def alert_severity(hazard: float, proximity: float, urgency: float, source_trust: float) -> float:
    weights = {"hazard": 0.35, "proximity": 0.2, "urgency": 0.3, "source_trust": 0.15}
    return round(
        clamp(
            hazard * weights["hazard"]
            + proximity * weights["proximity"]
            + urgency * weights["urgency"]
            + source_trust * weights["source_trust"]
        ),
        2,
    )


def security_risk(sensitivity: float, exposure: float, automation_power: float, source_trust: float) -> float:
    trust_penalty = 100.0 - clamp(source_trust)
    return round(
        clamp(sensitivity * 0.3 + exposure * 0.25 + automation_power * 0.25 + trust_penalty * 0.2),
        2,
    )


def queue_utilization(arrival_rate: float, service_rate: float) -> float:
    if service_rate <= 0:
        return 1.0
    return round(arrival_rate / service_rate, 4)


def exponential_smoothing(previous_forecast: float, actual: float, alpha: float = 0.35) -> float:
    bounded_alpha = max(0.0, min(1.0, alpha))
    return round(bounded_alpha * actual + (1 - bounded_alpha) * previous_forecast, 4)


def moat_score(
    network_effects: float,
    switching_cost: float,
    proprietary_data: float,
    brand: float,
    compliance: float,
    execution_speed: float,
) -> float:
    weights = {
        "network_effects": 0.18,
        "switching_cost": 0.2,
        "proprietary_data": 0.2,
        "brand": 0.14,
        "compliance": 0.16,
        "execution_speed": 0.12,
    }
    return round(
        clamp(
            network_effects * weights["network_effects"]
            + switching_cost * weights["switching_cost"]
            + proprietary_data * weights["proprietary_data"]
            + brand * weights["brand"]
            + compliance * weights["compliance"]
            + execution_speed * weights["execution_speed"]
        ),
        2,
    )


def pricing_score(cost_savings: float, time_savings: float, urgency: float, segment_fit: float) -> float:
    return round(clamp(cost_savings * 0.3 + time_savings * 0.25 + urgency * 0.2 + segment_fit * 0.25), 2)
